docs/source/rfc-configuration-overhaul.md

```diff
@@ -0,0 +1,721 @@
+
+- The message contains repetition.
+- Without an input snippet, the `Invalid character 's' at position 4` part of the message is not as helpful as it could be.
+
+#### 8. Config file with extra fields
+
+Providing a configuration file with extra fields that are not supposed to be there does not produce any errors at all.
+
+This might lead to a bad user experience when user expects some options to apply, but doesn't have any idea that those in fact are silently ignored.
+
+#### 9. Specifying `SUMERAGI.PUBLIC_KEY`
+
+As you can see in the "Redundant fields" section, user should not specify any non-null value for `SUMERAGI.PUBLIC_KEY` parameter. However, if they do, they will get the following error:
+
+```
+Error:
+   0: Please add `PUBLIC_KEY and PRIVATE_KEY` to the configuration.
+```
+
+The problem is that the error message tells something completely unrelated to the actual cause. This produces a terrible experience for users.
+
+#### 10. Config file is an invalid JSON file
+
+Providing an invalid JSON file as a config results in the same error we've seen before, which is not useful at all in this particular case and doesn't tell the user anything about the invalid JSON:
+
+```
+Error:
+   0: Please add `PUBLIC_KEY and PRIVATE_KEY` to the configuration.
+```
+
+### Ambiguity in Setting Root-Level Fields via ENV
+
+To understand the issue with ambiguity at the root-level fields, let's once again look at the excerpt from the [configuration reference](https://github.com/hyperledger/iroha/blob/35ba182e1d1b6b594712cf63bf448a2edefcf2cd/docs/source/references/config.md#network):
+
+> **`network`**
+>
+> Network configuration
+>
+> Has type `Option<network::ConfigurationProxy>`. Can be configured via environment variable `IROHA_NETWORK`
+>
+> ```json
+> {
+>   "ACTOR_CHANNEL_CAPACITY": 100
+> }
+> ```
+
+As we can see, `network` is not actually a configuration field in itself. Instead, there are various `network.*` fields that can be configured. This bring a question of what will happen when we set the `IROHA_NETWORK` environment variable as mentioned in the excerpt above? Will it override the nested `IROHA_NETWORK_CAPACITY` fields? Is there a use case for providing an ability to set `IROHA_NETWORK` all-in-one through an environment variable?
+
+### Chaotic Code Organisation
+
+Internally, not all of the configuration-related logic is contained within a single `iroha_config` crate. Instead, some of the configuration resolution logic is located in other crates, such as `iroha_cli`. This makes the configuration-related code error-prone and harder to maintain.
+
+## Proposals
+
+### Proposal 1 - Use TOML
+
+#### Objective
+
+Transition to TOML as Iroha's standard configuration format to provide a cleaner, more intuitive setup and to align with Rust community standards.
+
+#### Rationale
+
+- **Human-Friendly Syntax:** TOML's format is easier to read and understand compared to JSON, especially given its support for comments.
```

Comment:
If we want to adopt TOML we should probably cut down on nesting. In my opinion TOML doesn't handle it well

Reply:
Currently `wsv` has the deepest nesting:

```toml
[wsv]
assets_metadata_limits = {
  max_len = 1000,
  max_entry_byte_size = 32
}

# or

[wsv]
wasm_runtime_config.max_memory = 1_000_000

# or

[wsv.ident_length_limits]
min = 10
max = 32
```

I personally don't see any problem with that. The problems that come to my mind are:

- Readability: if nesting goes much deeper
- Familiarity: in comparison to JSON/YAML, TOML's syntax for nesting might be less intuitive

Could you expand your opinion?

Reply:
There are no formats that handle nesting well, with the possible exception of Nix. IMO the nesting we have is adequate for the format not to be confusing for users in TOML (or YAML).